[site-isolation] Add support for remote subframes with layerTreeAsText.

mattwoodrow opened the pull request:
#### c4ea9380f50ce075722cc8571170c9fbd7b755ee
<pre>
[site-isolation] Add support for remote subframes with layerTreeAsText.
<a href="https://bugs.webkit.org/show_bug.cgi?id=280727">https://bugs.webkit.org/show_bug.cgi?id=280727</a>
&lt;<a href="https://rdar.apple.com/137094708">rdar://137094708</a>&gt;

Reviewed by Said Abou-Hallawa.

Running layout tests with site-isolation enabled will require support for
layerTreeAsText including remote subframes.

This is largely the same as the renderTreeAsText implementation.

I don&apos;t think this change makes any tests pass yet, since there&apos;s still
differences between enabled/disabled. Future changes will hopefully improve
this.

It adds a new LayerTreeAsTextOption flag &apos;IncludeRootLayers&apos;. The default
behaviour is that the implicitly created &apos;root&apos; layers on the main frame&apos;s
RenderView&apos;s RenderLayer get skipped from the dump, but anything created when
recursing into subframes is included. This flag skips this, and gets added to
the flags passed between processes so that we don&apos;t skip root layer for non-main
root frames.

* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::dumpProperties const):
(WebCore::GraphicsLayer::layerTreeAsText const):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::layerTreeAsText):
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::dumpRemoteChildren const):
* Source/WebCore/platform/graphics/LayerTreeAsTextOptions.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::markFrontBufferVolatileForTesting):
(WebCore::RenderLayerBacking::dumpRemoteChildren const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::layerTreeAsText):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::renderTreeAsTextForTesting):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::renderTreeAsText):
(WebKit::WebRemoteFrameClient::layerTreeAsText):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::renderTreeAsTextForTesting):
(WebKit::WebPage::layerTreeAsTextForTesting):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/284653@main">https://commits.webkit.org/284653@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/400aef8073de35363232ccc0378a404d2f58c35f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70115 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22878 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/21281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72232 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21133 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/74202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/21281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73181 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45102 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60469 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41762 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17900 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/19650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63688 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18247 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75919 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14340 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17485 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14380 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60538 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/63267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15548 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11284 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4902 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45323 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/82 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46397 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47671 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46138 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
```diff
@@ -763,6 +763,7 @@ GenerateSyntheticEditingCommand(enum:uint8_t WebKit::SyntheticEditingCommandType
 
     RemotePostMessage(WebCore::FrameIdentifier source, String sourceOrigin, WebCore::FrameIdentifier target, std::optional<WebCore::SecurityOriginData> targetOrigin, struct WebCore::MessageWithMessagePorts message)
     RenderTreeAsTextForTesting(WebCore::FrameIdentifier frameID, size_t baseIndent, OptionSet<WebCore::RenderAsTextFlag> behavior) -> (String renderTreeDump) Synchronous
+   LayerTreeAsTextForTesting(WebCore::FrameIdentifier frameID, size_t baseIndent, OptionSet<WebCore::LayerTreeAsTextOptions> options) -> (String layerTreeDump) Synchronous
```
Please indent one more space.

shallawa commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15053,6 +15053,16 @@ void WebPageProxy::renderTreeAsTextForTesting(WebCore::FrameIdentifier frameID,
     completionHandler(WTFMove(result));
 }
 
+void WebPageProxy::layerTreeAsTextForTesting(WebCore::FrameIdentifier frameID, size_t baseIndent, OptionSet<WebCore::LayerTreeAsTextOptions> options, CompletionHandler<void(String&&)>&& completionHandler)
```
I think there is no need for adding the `WebCore` namespace here. At the beginning of this file there is
```
using namespace WebCore;
```

shallawa commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -2779,18 +2779,20 @@ void RenderLayerCompositor::updateCompositingForLayerTreeAsTextDump()
     updateCompositingForLayerTreeAsTextDump();
 
-    if (!m_rootContentsLayer)
+    RefPtr root = (options & LayerTreeAsTextOptions::IncludeRootLayers) ? rootGraphicsLayer() : m_rootContentsLayer;
```
This statement is hard to read. It looks `rootGraphicsLayer()` may return `m_overflowControlsHostLayer` which seems to be related to scrolling layers. Otherwise it returns `m_rootContentsLayer`. So I am not sure if `LayerTreeAsTextOptions::IncludeRootLayers` is the proper name here. Maybe `LayerTreeAsTextOptions::IncludeOverflowLayers` is a better name?

shallawa commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -2779,18 +2779,20 @@ void RenderLayerCompositor::updateCompositingForLayerTreeAsTextDump()
-    if (!m_rootContentsLayer)
+    RefPtr root = (options & LayerTreeAsTextOptions::IncludeRootLayers) ? rootGraphicsLayer() : m_rootContentsLayer;
+
+    if (!root)
```
root -> rootLayer?

shallawa commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -4425,4 +4426,14 @@ void RenderLayerBacking::markFrontBufferVolatileForTesting()
+        ts << indent << "(children "_s << 1 << "\n"_s;
+        ts << remoteFrame->client().layerTreeAsText(ts.indent() + 1, options);
+        ts << indent << ")\n"_s;
+
```
Please remove this extra line.

shallawa commented on Source/WebCore/platform/graphics/GraphicsLayer.cpp:
```diff
@@ -1074,7 +1074,8 @@ void GraphicsLayer::dumpProperties(TextStream& ts, OptionSet<LayerTreeAsTextOpti
         }
-    }
+    } else
+        client().dumpRemoteChildren(ts, options);
```
I think this statement is confusing. Does `GraphicsLayer` have to have either local children or remote children through its client? If the answer is yes then why I do not see any similar code path in this file for remote children?

This is why I do not like using `RemoteChildren` in the name. Why don't we call the new method all the times and without mentioning `Remote` in the name?
```
if (m_children.size()) {
    ....
}

client().dumpProperties(ts, options);
```

mattwoodrow replied:
With site-isolation, some GraphicsLayers have an identifier that references a GraphicsLayer child that lives in a different process. The actual parent-child connection only ever happens when the corresponding layer trees get pushed to the UI process (as RemoteLayerTreeNode/CALayer objects).

For normal runtime, GraphicsLayers only ever have local children, and the cross-process layer connections don't exist from the perspective of the WebProcess.

For testing, it's somewhat helpful to have `layerTreeAsText` work as if this weren't the case, and collect remote subtrees to include in the text dump. No other code in the WebProcess deals with these remote children (aside from assigning the identifiers that allow the connection to be made later).

Happy to push those details down into the client though.

shallawa commented on Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
```diff
@@ -95,6 +95,19 @@ String WebRemoteFrameClient::renderTreeAsText(size_t baseIndent, OptionSet<Rende
+    if (!page)
+        return "Test Error - Missing page"_s;
+    options.add(LayerTreeAsTextOptions::IncludeRootLayers);
+    auto sendResult = page->sendSync(Messages::WebPageProxy::LayerTreeAsTextForTesting(m_frame->frameID(), baseIndent, options));
```
I do not understand why do we have to have this complex workflow to dump the contents of the layers?

WebProcess: WebRemoteFrameClient::layerTreeAsText() -> sendMessage
UIProcess: WebPageProxy::layerTreeAsTextForTesting() -> sendMessage
WebProcess: WebPage::LayerTreeAsTextForTesting() -> completionHandler
UIProcess: WebPageProxy::layerTreeAsTextForTesting() ->completionHandler
WebProcess: WebRemoteFrameClient::layerTreeAsText()

mattwoodrow replied:
Is there another way?

The middle WebProcess in that list is a different process from the first/last, and we don't have direct WebProcess<->WebProcess IPC.

shallawa replied:
Oh I see. May be in the furniture we can implement a new paradigm to handle WebProcess<->WebProcess IPC via UIProcess without having to repeat the same code you did for `layerTreeAsText`

mattwoodrow committed:
[site-isolation] Add support for remote subframes with layerTreeAsText.
https://bugs.webkit.org/show_bug.cgi?id=280727
<rdar://137094708>

Reviewed by Said Abou-Hallawa.

Running layout tests with site-isolation enabled will require support for
layerTreeAsText including remote subframes.

This is largely the same as the renderTreeAsText implementation.

I don't think this change makes any tests pass yet, since there's still
differences between enabled/disabled. Future changes will hopefully improve
this.

It adds a new LayerTreeAsTextOption flag 'IncludeRootLayers'. The default
behaviour is that the implicitly created 'root' layers on the main frame's
RenderView's RenderLayer get skipped from the dump, but anything created when
recursing into subframes is included. This flag skips this, and gets added to
the flags passed between processes so that we don't skip root layer for non-main
root frames.

* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::dumpProperties const):
(WebCore::GraphicsLayer::layerTreeAsText const):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::layerTreeAsText):
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::dumpRemoteChildren const):
* Source/WebCore/platform/graphics/LayerTreeAsTextOptions.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::markFrontBufferVolatileForTesting):
(WebCore::RenderLayerBacking::dumpRemoteChildren const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::layerTreeAsText):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::renderTreeAsTextForTesting):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::renderTreeAsText):
(WebKit::WebRemoteFrameClient::layerTreeAsText):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::renderTreeAsTextForTesting):
(WebKit::WebPage::layerTreeAsTextForTesting):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/284653@main
